src/coreclr/System.Private.CoreLib/src/System/RuntimeHandles.cs

```diff
@@ -963,7 +964,18 @@ internal static MdUtf8String GetUtf8Name(RuntimeMethodHandleInternal method)
         [DebuggerStepThroughAttribute]
         [Diagnostics.DebuggerHidden]
         [MethodImpl(MethodImplOptions.InternalCall)]
-        internal static extern object InvokeMethod(object? target, object[]? arguments, Signature sig, bool constructor, bool wrapExceptions);
+        private static extern object? InvokeMethod(object? target, ref object? arguments, Signature sig, bool constructor, bool wrapExceptions);
+
+        internal static unsafe object? InvokeMethod(object? target, Span<object?> arguments, Signature sig, bool constructor, bool wrapExceptions)
+        {
+            // Native InvokeMethod needs the 'ref object' passed in to point to a pinned memory address (pinned array, stack-allocated
+            // struct, etc). We'll cast the object ref as a byte ref so that we can pin it, then we'll pass in the now-pinned object ref.
+            ref object? refToFirstArgElement = ref MemoryMarshal.GetReference(arguments);
+            fixed (byte* unused = &Unsafe.As<object?, byte>(ref refToFirstArgElement))
```

Comment:
It may be better to pass in the arguments into FCall as `ref Span<object> argument`. It would avoid the pinning  that is potentially very long (e.g. if entrypoint long-running operation is called via reflection).

Reply:
Is it legal to put an `OBJECTREFREF` or similar into the unmanaged GC struct? A quick search didn't turn up any other examples of this. If we could do this I think we can bypass the pinning like you suggest.

Reply:
Something like this: https://github.com/jkotas/runtime/commit/6fd1149d7149a1b4175333745c5ef2a45aa0a82c (untested, feel free to cherry pick the change).

Reply:
Just to make sure I understand this: you're passing the span byref, which means the "source of truth" for the span lives on the managed stack. When the GC runs, it'll make sure all those references are fixed up. So within the unmanaged runtime, as long as we always go back to this "source of truth" and never make our own local copy of the struct, we're good to go.

Does that accurately describe things?

Reply:
Yep. This is not the first place where FCalls/QCalls use this pattern. For example, `ObjectHandleOnStack` uses it too.

Reply:
Almost makes me want to keep `ByReference<T>` around so that we can pass it as an _in_ parameter via p/invoke. ;)